// strict-callbag-basics/api/auditTime.js
import { createPipe } from "./createPipe.js";
import { NONE } from "./none.js";
export const auditTimeP_ = (self, ms) => (_, sink) => {
    let lastItem = NONE;
    let timeout;
    const cleanup = () => {
        if (timeout) {
            clearTimeout(timeout);
            timeout = undefined;
        }
    };
    const maybeEmit = () => {
        if (lastItem !== NONE) {
            const data = lastItem;
            lastItem = NONE;
            sink(1, data);
        }
        timeout = undefined;
    };
    createPipe(self, sink, {
        onStart(s) {
            s.pull();
        },
        onRequest() {
            //
        },
        onData(s, data) {
            lastItem = data;
            if (!timeout) {
                timeout = setTimeout(maybeEmit, ms);
            }
            s.pull();
        },
        onEnd(err) {
            cleanup();
            maybeEmit();
            sink(2, err);
        },
        onAbort() {
            cleanup();
        },
    });
};
export const auditTimeP = (ms) => (self) => auditTimeP_(self, ms);
