// dfx/dist/DiscordConfig/index.js
import * as tsplus_module_1 from "dfx/common";
import * as tsplus_module_2 from "@fp-ts/data/Duration";
import * as tsplus_module_3 from "@effect/io/Layer";
import * as tsplus_module_4 from "@fp-ts/data/Function";
import * as tsplus_module_5 from "@effect/io/Effect";
const VERSION = 9;
export const DiscordConfig = tsplus_module_1.Tag();
export const make = ({ token, rest, gateway }) => ({
    token,
    rest: {
        baseUrl: `https://discord.com/api/v${VERSION}`,
        ...(rest ?? {}),
        globalRateLimit: {
            limit: 50,
            window: tsplus_module_2.seconds(1),
            ...(rest?.globalRateLimit ?? {}),
        },
    },
    gateway: {
        intents: tsplus_module_1.Discord.GatewayIntents.GUILDS,
        identifyRateLimit: [5000, 1],
        ...(gateway ?? {}),
    },
});
export const makeLayer = tsplus_module_4.flow(make, tsplus_module_3.succeed(DiscordConfig));
export const token = tsplus_module_5.serviceWith(DiscordConfig)(({ token }) => token);
export const rest = tsplus_module_5.serviceWith(DiscordConfig)(({ rest }) => rest);
export const gateway = tsplus_module_5.serviceWith(DiscordConfig)(({ gateway }) => gateway);
