// dfx/dist/Helpers/ui.js
import { ButtonStyle, ComponentType, TextInputStyle, } from "../types.js";
/**
 * Helper to create an Action Row grid.
 */
export const grid = (items) => items.map((components) => ({
    type: ComponentType.ACTION_ROW,
    components,
}));
/**
 * Helper to create a single column of components
 */
export const singleColumn = (items) => items.map((c) => ({
    type: ComponentType.ACTION_ROW,
    components: [c],
}));
/**
 * Helper to create a button component.
 */
export const button = (button) => ({
    type: ComponentType.BUTTON,
    style: ButtonStyle.PRIMARY,
    ...button,
});
/**
 * Helper to create a select component.
 */
export const select = (select) => ({
    type: ComponentType.STRING_SELECT,
    ...select,
});
/**
 * Helper to create a select component.
 */
export const userSelect = (select) => ({
    type: ComponentType.USER_SELECT,
    ...select,
});
/**
 * Helper to create a select component.
 */
export const roleSelect = (select) => ({
    type: ComponentType.ROLE_SELECT,
    ...select,
});
/**
 * Helper to create a select component.
 */
export const mentionableSelect = (select) => ({
    type: ComponentType.MENTIONABLE_SELECT,
    ...select,
});
/**
 * Helper to create a select component.
 */
export const channelSelect = (select) => ({
    type: ComponentType.CHANNEL_SELECT,
    ...select,
});
/**
 * Helper to create a text input
 */
export const textInput = (input) => ({
    type: ComponentType.TEXT_INPUT,
    style: TextInputStyle.SHORT,
    ...input,
});
