// dfx/dist/webhooks.js
import * as tsplus_module_1 from "dfx/common";
import * as tsplus_module_2 from "@effect/io/Layer";
export * as Discord from "./types.js";
export * as Config from "./DiscordConfig/index.js";
export { DiscordREST, LiveDiscordREST, rest } from "./DiscordREST/index.js";
export * as Ix from "./Interactions/index.js";
export * as Log from "./Log/index.js";
export * as RateLimitStore from "./RateLimitStore/index.js";
export * as Flags from "./Helpers/flags.js";
export * as IxHelpers from "./Helpers/interactions.js";
export * as Members from "./Helpers/members.js";
export * as Perms from "./Helpers/permissions.js";
export * as UI from "./Helpers/ui.js";
export const LiveRateLimit = tsplus_module_2.provideToAndMerge(tsplus_module_1.RateLimitStore.LiveRateLimiter)(tsplus_module_1.RateLimitStore.LiveMemoryRateLimitStore);
export const LiveREST = tsplus_module_2.provideToAndMerge(tsplus_module_1.Rest.LiveDiscordREST)(LiveRateLimit);
export const makeLayer = (config, debug = false) => {
    const LiveWebhook = tsplus_module_1.Ix.makeWebhookConfig(config);
    const LiveLog = debug ? tsplus_module_1.Log.LiveLogDebug : tsplus_module_1.Log.LiveLog;
    const LiveConfig = tsplus_module_1.Config.makeLayer(config);
    const LiveEnv = tsplus_module_2.provideToAndMerge(LiveREST)(tsplus_module_2.merge(LiveWebhook)(tsplus_module_2.merge(LiveConfig)(LiveLog)));
    return LiveEnv;
};
